src/backend/mpi/swdsm.cpp

```diff
@@ -541,291 +529,251 @@ void write_back_writebuffer() {
 }
 
 void load_cache_entry(unsigned long loadtag, unsigned long loadline) {
-	int i;
-	unsigned long homenode;
-	unsigned long id = 1 << getID();
-	unsigned long invid = ~id;
 
-	if(loadtag>=size_of_all){//Trying to access/prefetch out of memory
-		return;
+	if(loadtag >= size_of_all){ // Address out of bounds
+		printf("Address %lu out of bounds.\n", loadtag);
+		exit(EXIT_FAILURE);
 	}
-	homenode = getHomenode(loadtag);
-	unsigned long cacheIndex = loadline;
-	if(cacheIndex >= cachesize){
-		printf("idx > size   cacheIndex:%ld cachesize:%ld\n",cacheIndex,cachesize);
-		return;
+	if(loadline >= cachesize){ // Cache index out of bounds
+		printf("Cache index %ld out of bounds (cachesize: %lu).\n", loadline, cachesize);
+		exit(EXIT_FAILURE);
 	}
-	sem_wait(&ibsem);
-
 
-	unsigned long pageAddr = loadtag;
-	unsigned long blocksize = pagesize*CACHELINE;
-	unsigned long lineAddr = pageAddr/blocksize;
-	lineAddr *= blocksize;
+	/* Assign node bit IDs */
+	const unsigned long node_id_bit = 1 << getID();
+	const unsigned long node_id_inv_bit = ~node_id_bit;
 
-	unsigned long startidx = cacheIndex/CACHELINE;
-	startidx*=CACHELINE;
-	unsigned long end = startidx+CACHELINE;
+	/* Calculate start values and store some parameters */
+	const std::size_t block_size = pagesize*CACHELINE;
+	const unsigned long line_addr = (loadtag/block_size)*block_size;
+	const int start_index = (loadline/CACHELINE)*CACHELINE;
+	int end_index = start_index+CACHELINE;
+	const unsigned long load_node = getHomenode(line_addr);
+	bool have_lock = false;
 
-	if(end>=cachesize){
-		end = cachesize;
-	}
-
-	argo_byte tmpstate = cacheControl[startidx].state;
-	unsigned long tmptag = cacheControl[startidx].tag;
+	sem_wait(&ibsem);
 
-	if(tmptag == lineAddr && tmpstate != INVALID){
+	/* Return if cache entry is already up to date. */
+	if(cacheControl[start_index].tag == line_addr &&
+			cacheControl[start_index].state != INVALID){
 		sem_post(&ibsem);
 		return;
 	}
 
+	/* Adjust endidx to ensure the whole chunk to fetch is on the same node */
+	for(int j = start_index+CACHELINE, p = 1; static_cast<unsigned long>(j) <
```

Comment:
Stylistic comment: wouldn't this read better in 3 (one per "part" of the for) instead of 2 lines?

Reply:
Another comment related to this line (to `start_index` in particular), but it's a more general one, I noticed that the code uses `int` for some values (which should probably be `unsigned int`, really) in some places, and `std::size_t` in other places.

Reply:
Both of these suggestions should be reflected in the latest commit.